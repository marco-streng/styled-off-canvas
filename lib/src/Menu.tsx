import React from "react";
import styled from "styled-components";
import { useOffCanvas } from "./Context";

const Container = styled.div<{
  $isOpen: boolean;
  $background?: string;
  $duration?: string;
  $position?: "left" | "right";
  $width?: string;
}>(
  ({
    $isOpen,
    $background = "#fff",
    $duration = "500ms",
    $position = "right",
    $width = "300px",
  }) => `
  background-color: ${$background};
  height: 100%;
  left: ${$position === "left" ? ($isOpen ? 0 : "-100%") : "inherit"};
  max-width: ${$width};
  overflow-y: auto;
  position: fixed;
  right: ${$position === "right" ? ($isOpen ? 0 : "-100%") : "inherit"};
  top: 0;
  transition: ${$position} ${$duration} ease-in-out;
  width: 100%;
  z-index: 10;
`
);

const isCSR = typeof window !== "undefined";

const initalBodyStyle = {
  maxHeight: isCSR ? window.document.body.style.maxHeight : "none",
  overflow: isCSR ? window.document.body.style.overflow : "visbile",
};

/**
 * Menu component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
export const Menu = ({
  children,
  background,
  duration,
  position,
  width,
  closeOnEsc,
  className,
  ...props
}: {
  children: React.ReactNode;
  background?: string;
  duration?: string;
  position?: "left" | "right";
  width?: string;
  closeOnEsc?: boolean;
  className?: string;
}) => {
  const { isOpen, close } = useOffCanvas();

  // Avoid scrolling on content when the navigation is open
  React.useEffect(() => {
    const bodyElStyle = window.document.body.style;

    if (isOpen) {
      bodyElStyle.maxHeight = "100%";
      bodyElStyle.overflow = "hidden";
    } else {
      bodyElStyle.maxHeight = initalBodyStyle.maxHeight;
      bodyElStyle.overflow = initalBodyStyle.overflow;
    }
  }, [isOpen]);

  if (closeOnEsc) {
    React.useEffect(() => {
      const handleKeyDown = (event: KeyboardEvent) => {
        if (event.code.toLowerCase() === "escape") {
          close();
        }
      };

      isOpen
        ? window.document.addEventListener("keydown", handleKeyDown)
        : window.document.removeEventListener("keydown", handleKeyDown);
    }, [isOpen]);
  }

  return (
    <Container
      className={className}
      $background={background}
      $duration={duration}
      $isOpen={isOpen}
      $position={position}
      $width={width}
      {...props}
    >
      {children}
    </Container>
  );
};
