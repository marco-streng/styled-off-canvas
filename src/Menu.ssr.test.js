/**
 * @jest-environment node
 */

import React from 'react'
import TestRenderer from 'react-test-renderer'

import Provider from './Provider'
import Menu from './Menu'

describe('Menu', () => {
  let useEffectSpy

  beforeEach(() => {
    // useEffect won't run on server
    useEffectSpy = jest.spyOn(React, 'useEffect')
    useEffectSpy.mockImplementation(() => jest.fn())
  })

  afterEach(() => {
    useEffectSpy.mockRestore()
  })

  test('ssr', () => {
    const tree = TestRenderer.create(
      <Provider onClose={() => {}} isOpen><Menu><div>content</div></Menu></Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
