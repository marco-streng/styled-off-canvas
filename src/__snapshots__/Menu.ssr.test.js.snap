// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu ssr 1`] = `
<div
  background="#fff"
  className="Menu__Container-sc-mw2r80-0 hMKqmI"
  duration="500ms"
  isOpen={true}
  position="right"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;
